### PR TITLE
fix: json import balances for most chains never loading during onboarding

### PR DESCRIPTION
**Pull request by alecdwm**
- Fixed the backend subscription so that it closes properly & so that closing subscriptions never send old data to the frontend
- Fixed the frontend incorrectly formatting the addresses used as params using chain-specific prefixes. No other balances code does this, because it makes it impossible to then filter the balances in the frontend.

### CODE REVIEW

File: apps/extension/src/ui/hooks/useAccountImportBalances.tsx [hunk 1]
```diff
@@ -33,7 +31,7 @@ export const useAccountImportBalances = (accounts: AccountImportDef[]) => {
     const addressesByChain: AddressesByChain = chains.reduce((prev, chain) => {
       const addresses = safeAccounts
         .filter(({ type, genesisHash }) => isAccountCompatibleWithChain(chain, type, genesisHash))
-        .map((account) => convertAddress(account.address, chain.prefix))
```
Comment (alecdwm) [inline apps/extension/src/ui/hooks/useAccountImportBalances.tsx:36]: the source of the bug

File: apps/extension/src/core/domains/balances/handler.ts [hunk 2]
```diff
@@ -89,107 +89,101 @@ const subscribeBalancesByParams = async (
...
     miniMetadataIds: [],
-  })
+  }
+  let balancesSubscriptionsGeneration = 0
+  let balancesUnsubCallbacks: Promise<UnsubscribeFn>[] = []
+
+  // create a subscription to the balances params
+  // allows us to restart the inner balances subscriptions whenever the data
+  // which they depend on changes
+  const byParamsSubscription = combineLatest([
```
Comment (alecdwm) [inline apps/extension/src/core/domains/balances/handler.ts:106]: This was never closed before

File: apps/extension/src/core/domains/balances/handler.ts [hunk 3]
```diff
@@ -89,107 +89,101 @@ const subscribeBalancesByParams = async (
...
+          (error, result) => {
+            if (thisGeneration !== balancesSubscriptionsGeneration) return
+
+            // eslint-disable-next-line no-console
+            if (error) return DEBUG && console.error(error)
 
-  // restart subscriptions each type params update
-  obsSubscriptionParams.subscribe(async ({ addressesByTokenByModule }) => {
-    // close previous subscriptions
-    await Promise.all(closeSubscriptionCallbacks)
```
Comment (alecdwm) [inline apps/extension/src/core/domains/balances/handler.ts:151]: This didn't actually close them

File: apps/extension/src/core/domains/balances/handler.ts [hunk 4]
```diff
@@ -89,107 +89,101 @@ const subscribeBalancesByParams = async (
...
+    chaindataProvider.evmNetworksListObservable,
+    // tokens
+    chaindataProvider.tokensListObservable,
+    // miniMetadatas - not used here but we must retrigger the subscriptions when this changes
+    liveQuery(async () => await balancesDb.miniMetadatas.toArray()),
+
+    // active state of substrate chains
+    activeChainsStore.observable,
+    // active state of evm networks
+    activeEvmNetworksStore.observable,
```
Comment (alecdwm) [inline apps/extension/src/core/domains/balances/handler.ts:119]: By changing the order of these, they can be passed into `getSubscriptionParams` just by doing `...args`

File: apps/extension/src/core/domains/balances/handler.ts [hunk 5]
```diff
@@ -89,107 +89,101 @@ const subscribeBalancesByParams = async (
...
-          addressesAndEvmNetworks,
-          addressesAndTokens,
-          chains,
-          evmNetworks,
-          tokens,
-          activeChains,
-          activeEvmNetworks,
-          activeTokens,
-          miniMetadatas
+      const thisGeneration = balancesSubscriptionsGeneration
```
Comment (alecdwm) [inline apps/extension/src/core/domains/balances/handler.ts:140]: Prevent async updates from closing-but-not-yet-closed subscriptions from being transmitted through to the frontend

File: apps/extension/src/core/domains/balances/handler.ts [hunk 6]
```diff
@@ -89,107 +89,101 @@ const subscribeBalancesByParams = async (
...
-    )
+    },
   })
 
   // unsub on port disconnect
   port.onDisconnect.addListener((): void => {
     unsubscribe(id)
-    closeSubscriptionCallbacks.forEach((cb) => cb.then((close) => close()))
+    byParamsSubscription.unsubscribe()
+    balancesUnsubCallbacks.forEach((cb) => cb.then((close) => close()))
```
Comment (alecdwm) [inline apps/extension/src/core/domains/balances/handler.ts:186]: Unsub all subscriptions (byParams as well as the inner balances subs) when the UI requesting these balances is closed

**Review by chidg — APPROVED** [review]
LGTM!

**Pull request merged**